modules/foreign-state-oracle/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+
+// Origin with arbitrary bytes included
+#[derive(PartialEq, Eq, Clone, RuntimeDebug, Encode, Decode, TypeInfo)]
+pub struct RawOrigin {
+	data: Vec<u8>,
+}
+
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct VerifiableCall<Call, BlockNumber, Balance> {
+	// Call to be dispatched by oracle
+	dispatchable_call: Call,
+	// Blocknumber at which call will be expired
+	expiry: BlockNumber,
+	// Reward available for responding to this query
+	oracle_reward: Balance,
+}
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	pub type VerifiableCallOf<T> =
+		VerifiableCall<<T as Config>::VerifiableTask, <T as frame_system::Config>::BlockNumber, Balance>;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The outer origin type.
+		type Origin: From<RawOrigin>;
+
+		/// Currency for query payments
+		type Currency: NamedReservableCurrency<
+			Self::AccountId,
+			Balance = Balance,
+			ReserveIdentifier = ReserveIdentifier,
+		>;
+
+		/// The foreign state oracle module id, keeps expired queries deposits
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Fee to be paid to oracles for servicing query
+		#[pallet::constant]
+		type QueryFee: Get<Balance>;
+
+		/// Fee for cancelling query
+		#[pallet::constant]
+		type CancelFee: Get<Balance>;
+
+		/// Timeout for query requests
+		#[pallet::constant]
+		type QueryDuration: Get<Self::BlockNumber>;
+
+		/// Dispatchable task that needs to be verified by oracle for dispatch
+		type VerifiableTask: Parameter
+			+ Dispatchable<Origin = <Self as Config>::Origin, PostInfo = PostDispatchInfo>
+			+ From<frame_system::Call<Self>>
+			+ GetDispatchInfo;
+
+		/// Origin that can dispatch calls that have been verified with foreign state
+		type OracleOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+
+		/// Provides current blocknumber
+		type BlockNumberProvider: BlockNumberProvider<BlockNumber = Self::BlockNumber>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::origin]
+	pub type Origin = RawOrigin;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Active Query is created, under the index as the key
+		CreateActiveQuery { index: QueryIndex, expiry: T::BlockNumber },
+		/// Call is dispatched with data, includes the result of the extrinsic
+		CallDispatched { task_result: DispatchResult },
+		/// Query that has expired is removed from storage, includes index
+		StaleQueryRemoved { query_index: QueryIndex },
+		/// Query is canceled, includes index
+		QueryCanceled { index: QueryIndex },
+	}
+
+	/// Index of Queries, each query gets unique number
+	#[pallet::storage]
+	#[pallet::getter(fn query_index)]
+	pub(super) type QueryCounter<T: Config> = StorageValue<_, QueryIndex, ValueQuery>;
```

Comment:
why getter is index but storage is counter? also better call it `NextQueryId` to be consistent with other pallets

Reply:
Renamed